src/runtime/profiler_common.cpp

```diff
@@ -272,34 +328,34 @@ WEAK void halide_profiler_memory_allocate(void *user_context,
         return;
     }
 
-    halide_profiler_pipeline_stats *p_stats = (halide_profiler_pipeline_stats *)pipeline_state;
-    halide_abort_if_false(user_context, p_stats != nullptr);
+    halide_abort_if_false(user_context, instance != nullptr);
     halide_abort_if_false(user_context, func_id >= 0);
-    halide_abort_if_false(user_context, func_id < p_stats->num_funcs);
+    halide_abort_if_false(user_context, func_id < instance->pipeline_stats->num_funcs);
 
-    halide_profiler_func_stats *f_stats = &p_stats->funcs[func_id];
+    halide_profiler_func_stats *func = &instance->funcs[func_id];
 
     // Note: Update to the counter is done without grabbing the state's lock to
-    // reduce lock contention. One potential issue is that other call that frees the
-    // pipeline and function stats structs may be running in parallel. However, the
-    // current desctructor (called on profiler shutdown) does not free the structs
-    // unless user specifically calls halide_profiler_reset().
-
-    // Update per-pipeline memory stats
-    atomic_add_fetch_sequentially_consistent(&p_stats->num_allocs, 1);
-    atomic_add_fetch_sequentially_consistent(&p_stats->memory_total, incr);
-    uint64_t p_mem_current = atomic_add_fetch_sequentially_consistent(&p_stats->memory_current, incr);
-    sync_compare_max_and_swap(&p_stats->memory_peak, p_mem_current);
+    // reduce lock contention. One potential issue is that another call that
+    // frees the pipeline and function stats structs may be running in
+    // parallel. However, the current destructor (called on profiler shutdown)
+    // does not free the structs unless user specifically calls
+    // halide_profiler_reset().
+
+    // Update per-instance memory stats
+    atomic_add_fetch_sequentially_consistent(&instance->num_allocs, 1);
+    atomic_add_fetch_sequentially_consistent(&instance->memory_total, incr);
```

Comment:
Same question: atomics because of parallelism within one pipeline invocation?

Reply:
Yes, these variables are hammered by multiple threads within a single invocation if there's a parallel loop.